frontend/src/metabase/visualizations/types/visualization.ts

```diff
@@ -176,24 +193,59 @@ export type VisualizationSettingDefinition<TValue, TProps = void> = {
   getMarginBottom?: (series: Series, settings: VisualizationSettings) => string;
   persistDefault?: boolean;
   inline?: boolean;
-  props?: TProps;
-  getProps?: (
-    series: Series,
-    vizSettings: VisualizationSettings,
-    onChange: (value: TValue) => void,
-    extra: unknown,
-    onChangeSettings: (value: Record<string, any>) => void,
-  ) => TProps;
   readDependencies?: string[];
   writeDependencies?: string[];
   eraseDependencies?: string[];
-  // is the setting visible in the dashboard card viz settings
   dashboard?: boolean;
   useRawSeries?: boolean;
+  hidden?: boolean;
+  placeholder?: string;
+  index?: number;
 };
 
+export type WidgetComponentDef<Widget, P, TValue> = {
+  widget?: Widget;
+  props?: P;
+  getProps?: (
+    series: Series,
+    vizSettings: VisualizationSettings,
+    onChange: (value: TValue | null) => void,
+    extra: SettingDefinitionExtra,
+    onChangeSettings: (value: VisualizationSettings) => void,
+  ) => P;
+};
+
+export type WidgetStringDef<Widget, TValue> = {
+  widget?: Widget;
+  props?: unknown;
+  getProps?: (
+    series: Series,
+    vizSettings: VisualizationSettings,
+    onChange: (value: TValue | null) => void,
+    extra: SettingDefinitionExtra,
+    onChangeSettings: (value: VisualizationSettings) => void,
+  ) => unknown;
+};
+
+type VisualizationSettingWidgetDef<
+  TValue,
+  Widget extends string | ComponentType<any> = string | ComponentType<any>,
+> = Widget extends string
+  ? WidgetStringDef<Widget, TValue>
+  : Widget extends ComponentType<infer P>
+    ? WidgetComponentDef<Widget, P, TValue>
+    : never;
```

Comment:
I wonder if we could unify `WidgetStringDef` and `WidgetComponentDef` here, and use `WidgetDef<Widget, unknown, Value>` for the string definition, as the only difference between the two is the Props being unknown. I think this is easier to read though, so no need to change this.

Another note is we could use `ComponentType<infer Props>` to make the type parameter a bit explicit